File: app/src/main/java/com/example/win/easy/repository/db/dao/SongListDao.java
package com.example.win.easy.repository.db.dao;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import com.example.win.easy.repository.db.data_object.SongListDO;

import java.util.Collection;
import java.util.List;

@Dao
public interface SongListDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Long insert(SongListDO songListDO);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    Long[] insert(Collection<SongListDO> songListDOs);

    //以下为Repo涉及方法
    @Query("SELECT * FROM SongListDO")
    List<SongListDO> getAllSongList();

    @Query("SELECT * FROM SongListDO WHERE remoteId=:songListRemoteId")
    SongListDO findByRemoteId(Long songListRemoteId);

    @Query("SELECT * FROM SongListDO WHERE remoteId is not null")
    List<SongListDO> findAllDataOnWeb();

    @Query("SELECT * FROM SongListDO WHERE remoteId is null")
    List<SongListDO> findAllDataOnLocal();
    /***********************/


    @Query("SELECT * FROM SongListDO")
    LiveData<List<SongListDO>> findAllSongListDOs();

    @Query("SELECT * FROM SongListDO WHERE id=:songListId")
//    SongListDO findById(Long songListId);
    LiveData<SongListDO> findById(Long songListId);

    @Query("SELECT * FROM SongListDO WHERE source=:dataSource")
    List<SongListDO> findByDataSource(String dataSource);
//    LiveData<List<SongListDO>> findByDataSource(String dataSource);

    @Query("SELECT * FROM SongListDO WHERE name=:name AND source=:source")
    List<SongListDO> findAllByNameAndSource(String name, String source);

    @Delete
    void delete(SongListDO songListDO);

    @Delete
    void delete(Collection<SongListDO> songListDOs);

    @Update
    void update(SongListDO songListDO);

    @Update
    void update(Collection<SongListDO> songListDOs);
}

File: app/src/main/java/com/example/win/easy/repository/db/dao/SongListDao.kt
package com.example.win.easy.repository.db.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.example.win.easy.repository.db.data_object.SongDO
import com.example.win.easy.repository.db.data_object.SongListDO

@Dao
interface SongListDao {
    @Query("SELECT * FROM SongListDO")
    fun loadAll(): LiveData<List<SongListDO>>

    @Query("SELECT * FROM SongListDO join SongXSongListDO WHERE  songId=:songId")
    suspend fun loadBySong(songId: Long): List<SongListDO>

    @Insert(onConflict = OnConflictStrategy.ABORT)
    suspend fun insert(songListDO: SongListDO): Long

    //以下为Repo涉及方法
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insert(songListDOs: Collection<SongListDO?>?): Array<Long?>?

    @Query("SELECT * FROM SongListDO")
    fun allSongList(): List<SongListDO?>?

    @Query("SELECT * FROM SongListDO WHERE remoteId=:songListRemoteId")
    fun findByRemoteId(songListRemoteId: Long?): SongListDO?

    @Query("SELECT * FROM SongListDO WHERE remoteId is not null")
    fun findAllDataOnWeb(): List<SongListDO?>?

    @Query("SELECT * FROM SongListDO WHERE remoteId is null")
    fun findAllDataOnLocal(): List<SongListDO?>?

    @Query("SELECT * FROM SongDO join SongXSongListDO WHERE songListId=:songListId")
    fun findSongsInSongList(songListId: Long): LiveData<List<SongDO>?>

    /** */
    @Query("SELECT * FROM SongListDO WHERE id=:songListId")
    fun  //    SongListDO findById(Long songListId);
            findById(songListId: Long?): LiveData<SongListDO?>?

    @Query("SELECT * FROM SongListDO WHERE source=:dataSource")
    fun findByDataSource(dataSource: String?): List<SongListDO?>?

    //    LiveData<List<SongListDO>> findByDataSource(String dataSource);
    @Query("SELECT * FROM SongListDO WHERE name=:name AND source=:source")
    fun findAllByNameAndSource(name: String?, source: String?): List<SongListDO?>?

    @Delete
    fun delete(songListDO: SongListDO?)

    @Delete
    fun delete(songListDOs: Collection<SongListDO?>?)

    @Update
    fun update(songListDO: SongListDO?)

    @Update
    fun update(songListDOs: Collection<SongListDO?>?)
}